plugins/catalog-backend/src/ingestion/processors/microsoftGraph/client.ts

```diff
@@ -0,0 +1,232 @@
+    const photos = result.value as MicrosoftGraph.ProfilePhoto[];
+    let selectedPhoto: MicrosoftGraph.ProfilePhoto | undefined = undefined;
+
+    // Find the biggest picture that is small than the max size
+    for (const p of photos) {
+      if (
+        !selectedPhoto ||
+        (p.height! >= selectedPhoto.height! && p.height! <= maxSize)
+      ) {
+        selectedPhoto = p;
+      }
+    }
+
+    if (!selectedPhoto) {
+      return undefined;
+    }
+
+    return await this.getUserPhoto(userId, selectedPhoto.id!);
+  }
+
+  async getUserPhoto(
+    userId: string,
+    sizeId?: string,
+  ): Promise<string | undefined> {
+    const path = sizeId
+      ? `users/${userId}/photos/${sizeId}/$value`
+      : `users/${userId}/photo/$value`;
+    const response = await this.api(path);
+
+    if (response.status === 404) {
+      return undefined;
+    } else if (response.status !== 200) {
+      await this.handleError('photo', response);
+    }
+
+    return `data:image/jpeg;base64,${Buffer.from(
+      await response.arrayBuffer(),
+    ).toString('base64')}`;
+  }
+
+  async *getUsers(query?: ODataQuery): AsyncIterable<MicrosoftGraph.User> {
+    let response = await this.api(`users`, query);
+    for (;;) {
+      if (response.status !== 200) {
+        await this.handleError('users', response);
+      }
+
+      const result = await response.json();
+      const users: [MicrosoftGraph.User] = result.value;
+
+      yield* users;
+
+      // Follow cursor to the next page if one is available
+      if (!result['@odata.nextLink']) {
+        return;
+      }
+
+      response = await this.request(result['@odata.nextLink']);
+    }
+  }
+
+  async *getGroups(query?: ODataQuery): AsyncIterable<MicrosoftGraph.Group> {
+    let response = await this.api(`groups`, query);
+
+    for (;;) {
+      if (response.status !== 200) {
+        await this.handleError('groups', response);
+      }
+
+      const result = await response.json();
+      const groups: [MicrosoftGraph.Group] = result.value;
+
+      yield* groups;
+
+      // Follow cursor to the next page if one is available
+      if (!result['@odata.nextLink']) {
+        return;
+      }
+
+      if (!result['@odata.nextLink']) {
+        return;
+      }
+
+      response = await this.request(result['@odata.nextLink']);
+    }
+  }
+
+  async *getGroupMembers(
+    groupId: string,
+  ): AsyncIterable<
+    | (MicrosoftGraph.Group & { '@odata.type': '#microsoft.graph.user' })
+    | (MicrosoftGraph.User & { '@odata.type': '#microsoft.graph.group' })
+  > {
+    let response = await this.api(`groups/${groupId}/members`);
+
+    for (;;) {
+      if (response.status !== 200) {
+        await this.handleError('group members', response);
+      }
+
+      const result = await response.json();
+      const groups: [] = result.value;
+
+      yield* groups;
+
+      // Follow cursor to the next page if one is available
+      if (!result['@odata.nextLink']) {
```

Comment:
This whole "iterate through an odata list" stuff can probably be broken out into a helper